mono/metadata/boehm-gc.c

```diff
@@ -92,27 +92,45 @@ void mono_gc_handle_lock () { lock_handles (NULL);}
 #define unlock_handles(handles) mono_os_mutex_unlock (&handle_section)
 void mono_gc_handle_unlock () { unlock_handles (NULL); }
 
-typedef struct {
+/*
+ * The value of HANDLE_COUNT must satisfy two constraints:
+ * 1. It must be a multiple of CHAR_BIT, e.g. HANDLE_COUNT % CHAR_BIT == 0
+ * 2. It must be small enough that sizeof(HandleData) < HANDLE_DATA_ALIGNMENT
+ *
+ * The two values chosen below for 32 & 64 bit, are the largest values for HANDLE_COUNT
+ * to result in the smallest amount of wasted space at the end of the allocation.
+ */
+#if SIZEOF_VOID_P == 4
+#define HANDLE_COUNT 1008
+#define HANDLE_DATA_ALIGNMENT 4096
+#else
+#define HANDLE_COUNT 992
+#define HANDLE_DATA_ALIGNMENT 8192
+#endif
+typedef struct HandleData HandleData;
+struct HandleData {
+	HandleData *next; //immutable
+	HandleData *next_free; // next free
 	guint32  *bitmap;
-	gpointer *entries;
-	guint32   size;
-	guint8    type;
-	guint     slot_hint : 24; /* starting slot for search in bitmap */
 	/* 2^16 appdomains should be enough for everyone (though I know I'll regret this in 20 years) */
 	/* we alloc this only for weak refs, since we can get the domain directly in the other cases */
 	guint16  *domain_ids;
-} HandleData;
+	guint32   in_use;
+	guint32   size;
+	guint8    type;
+	guint     slot_hint : 24; /* starting slot for search in bitmap */
+	gpointer  entries[HANDLE_COUNT];
+};
 
-#define EMPTY_HANDLE_DATA(type) {NULL, NULL, 0, (type), 0, NULL}
 
-/* weak and weak-track arrays will be allocated in malloc memory 
- */
-static HandleData gc_handles [] = {
-	EMPTY_HANDLE_DATA (HANDLE_WEAK),
-	EMPTY_HANDLE_DATA (HANDLE_WEAK_TRACK),
-	EMPTY_HANDLE_DATA (HANDLE_NORMAL),
-	EMPTY_HANDLE_DATA (HANDLE_PINNED)
-};
+#define MONO_GC_HANDLE_TO_UINTPTR(ptr) ((uintptr_t)((size_t)(ptr)))
+
+#define GC_HANDLE_TYPE_IS_WEAK(x) ((x) <= HANDLE_WEAK_TRACK)
+
+#define MONO_GC_HANDLE_TYPE_IS_WEAK(x) ((x) <= HANDLE_WEAK_TRACK)
+
+static HandleData* gc_handles[4];
+static HandleData* gc_handles_free[4];
```

Comment:
How are these initialized to zero?  The first call to `alloc_handle` checks for null. Not sure where these are zeroed.

Reply:
All statics are zero initialized by the loader in C/C++.